overlay.go

```diff
@@ -331,3 +335,321 @@ func isZero(v reflect.Value) bool {
 		panic(fmt.Sprintf("unexpected value of type %s passed to isZero", v.Kind().String()))
 	}
 }
+
+// ReadAndMergeBundleData reads N bundle data sources, composes their contents
+// together and returns the result. The first bundle data source is treated as
+// a base bundle while subsequent bundle data sources are treated as overlays
+// which are sequentially merged onto the base bundle.
+//
+// Before returning the merged bundle, ReadAndMergeBundleData will also attempt
+// to resolve any include directives present in the machine annotations,
+// application options and annotations by invoking the provided includeResolver
+// with the path to the file required for each include.
+//
+// When merging an overlay into a base bundle the following rules apply for the
+// BundleData struct fields:
+// - if an overlay specifies a bundle-level series, it overrides the base bundle
+//   series.
+// - overlay-defined relations are appended to the base bundle relations
+// - overlay-defined machines overwrite the baase bundle machines.
+// - if an overlay defines an application that is not present in the base bundle,
+//   it will get appended to the application list.
+// - if an overlay defines an empty/nil application, it will be removed from the
+//   base bundle together with any associated relations.
+// - if an overlay defines an application that is also present in the base bundle
+//   the two application specs are merged together (see following rules)
+//
+// ApplicationSpec merge rules:
+// - if the overlay defines a value for a scalar or slice field, it will overwrite
+//   the value from the base spec (e.g. trust, series etc).
+// - if the overlay specifies a nil/empty value for a map field, then the map
+//   field of the base spec will be cleared.
+// - if the overlay specifies a non-empty value for a map field, its key/value
+//   tuples are iterated and:
+//   - if the value is nil/zero, it is deleted from the base spec
+//   - otherwise, the key/value is inserted into the base spec overwriting any
+//     existing entries.
+func ReadAndMergeBundleData(includeResolver func(path string) ([]byte, error), sources ...BundleDataSource) (*BundleData, error) {
+	var allParts []*bundleDataPart
+	for _, src := range sources {
+		parts, err := src.read()
+		if err != nil {
+			return nil, err
+		}
+
+		allParts = append(allParts, parts...)
+	}
+
+	if len(allParts) == 0 {
+		return nil, errors.New("malformed bundle: bundle has no content")
+	}
+
+	// Treat allParts[0] as the base bundle and all other parts as overlays.
+	base := allParts[0]
+	if err := VerifyNoOverlayFieldsPresent(base.data); err != nil {
+		return nil, err
+	}
+
+	for _, overlay := range allParts[1:] {
+		if err := applyOverlay(base, overlay); err != nil {
+			return nil, err
+		}
+	}
+
+	if includeResolver == nil {
+		return base.data, nil
+	}
+
+	for app, appData := range base.data.Applications {
```

Comment:
is it ok to resolve includes after applying overlays? I would hope thats ok, just checking as I think it is a change in logic.
(previously I think you couldn't put includes in overlays?)

Reply:
This behavior matches the current implementation: https://github.com/juju/juju/blob/develop/cmd/juju/application/bundle.go#L76

However, the current implementation _seems to assume_ that only the base bundle may contain include directives as the code resolves include paths relative to the base bundle path (https://github.com/juju/juju/blob/develop/cmd/juju/application/bundle.go#L143) which kinda limits the usefulness of having includes in overlays.